lib/rubocop/cop/lint/nested_percent_literal.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks for nested percent literals.
+      #
+      # @example
+      #
+      #   # bad
+      #
+      #   # The percent literal for nested_attributes is parsed as four tokens,
+      #   # yielding the array [:name, :content, :"%i[incorrectly", :"nested]"].
+      #   attributes = {
+      #     valid_attributes: %i[name content],
+      #     nested_attributes: %i[name content %i[incorrectly nested]]
+      #   }
+      class NestedPercentLiteral < Cop
+        include PercentLiteral
+
+        MSG = 'Do not nest percent literals.'.freeze
+
+        # The array of regular expressions representing percent literals that,
+        # if found within a percent literal expression, will cause a
+        # NestedPercentLiteral violation to be emitted.
+        REGEXES = PercentLiteral::PERCENT_LITERAL_TYPES.map do |percent_literal|
+          /#{percent_literal}\W/
+        end.freeze
+
+        def on_array(node)
+          process(node, *PercentLiteral::PERCENT_LITERAL_TYPES)
+        end
+
+        def on_percent_literal(node)
+          add_offense(node) if contains_percent_literals?(node)
+        end
+
+        private
+
+        # Returns whether node contains any tokens that match a percent literal
+        # invokation.
+        def contains_percent_literals?(node)
+          node.children.any? do |child|
```

Comment:
Check the `ArrayNode` node extension. There should be a method to check if a node is a percent literal. Then your check can be something like:

```ruby
node.each_child(:array).any? { |child| child.percent_literal? }
```

Reply:
This doesn't seem to work (at the very least, as written, offenses no longer get emitted), presumably because the AST parses the nested percent literals differently (being tokens to the outer nested percent literal). See the TODO in the last test, I think it is related.

Reply:
Alas. I forgot the parser does this. Check if there's a `Node#str_content` method though. 🙂

Reply:
There is. But it isn't clear to me how it makes sense to use it. Converting the tokens into strings, joining them, and reparsing them doesn't seem appropriate. Is this what you were thinking?